Improve error message with unsupported cpp

### Description

Currently, if you have a `cpp` that's too old you get an error like this when compiling:
```
src/battle_main.c:338:84: error: expected expression before ',' token
  338 |     TRAINER_CLASS(HIKER, "HIKER", 10),
```
This has come up a few times on Discord, [most recently today](<https://discord.com/channels/419213663107416084/774393519569502268/1208339685471158273>). It's not at all clear what the problem is, especially because the error message doesn't even say `__VA_OPT__`.

We could probably give a better error message by detecting [`__VA_OPT__` support](https://stackoverflow.com/questions/48045470/portably-detect-va-opt-support) in [`metaprogram.h`](https://github.com/rh-hideout/pokeemerald-expansion/blob/upcoming/include/metaprogram.h), and using [`#error`](https://en.cppreference.com/w/cpp/preprocessor/error) to give an explicit message rather than this very weird and confusing Hiker one.

### Version

upcoming (Edge)

### Upcoming/master Version

_No response_

### Discord contact info

_No response_

--- a/include/metaprogram.h
+++ b/include/metaprogram.h
@@ -2,6 +2,15 @@
 #ifndef METAPROGRAM_H
 #define METAPROGRAM_H
 
+/* Check if VA_OPT_ is supported by the compiler. GCC's version should be at least 9.5*/
+#define PP_THIRD_ARG(a,b,c,...) c
+#define VA_OPT_SUPPORTED_I(...) PP_THIRD_ARG(__VA_OPT__(,),true,false,)
+#define VA_OPT_SUPPORTED VA_OPT_SUPPORTED_I(?)
+
+#if !VA_OPT_SUPPORTED
+#error ERROR: VA_OPT__ is not supported. Please update your gcc compiler to version 10 or higher
+#endif // VA_OPT_SUPPORTED
+
 /* Calls m0/m1/.../m8 depending on how many arguments are passed. */
 #define VARARG_8(m, ...) CAT(m, NARG_8(__VA_ARGS__))(__VA_ARGS__)
 

--- a/include/metaprogram.h
+++ b/include/metaprogram.h
@@ -4,7 +4,7 @@
 
 /* Check if VA_OPT_ is supported by the compiler. GCC's version should be at least 9.5*/
 #define PP_THIRD_ARG(a,b,c,...) c
-#define VA_OPT_SUPPORTED_I(...) PP_THIRD_ARG(__VA_OPT__(,),true,false,)
+#define VA_OPT_SUPPORTED_I(...) PP_THIRD_ARG(__VA_OPT__(,),TRUE,FALSE,)
 #define VA_OPT_SUPPORTED VA_OPT_SUPPORTED_I(?)
 
 #if !VA_OPT_SUPPORTED

